bulkcopy.go

```diff
@@ -0,0 +1,576 @@
+	Options    MssqlBulkOptions
+	Debug      bool
+}
+type MssqlBulkOptions struct {
+	CheckConstraints  bool
+	FireTriggers      bool
+	KeepNulls         bool
+	KilobytesPerBatch int
+	RowsPerBatch      int
+	Order             []string
+	Tablock           bool
+}
+
+type DataValue interface{}
+
+func (cn *MssqlConn) CreateBulk(table string, columns []string) (_ *MssqlBulk) {
+	b := MssqlBulk{cn: cn, tablename: table, headerSent: false, columnsName: columns}
+	b.Debug = false
+	return &b
+}
+
+func (b *MssqlBulk) sendBulkCommand() error {
+
+	//get table columns info
+	err := b.getMetadata()
+	if err != nil {
+		return err
+	}
+	//match the columns
+	if b.Debug {
+		log.Printf("columnsName: %s", b.columnsName)
+	}
+	for _, colname := range b.columnsName {
+		var bulkCol *columnStruct
+
+		for _, m := range b.metadata {
+			if strings.EqualFold(m.ColName, colname) {
+				bulkCol = &m
+				break
+			}
+		}
+		if bulkCol != nil {
+
+			if bulkCol.ti.TypeId == typeUdt {
+				//send udt as binary
+				bulkCol.ti.TypeId = typeBigVarBin
+			}
+			b.bulkColumns = append(b.bulkColumns, *bulkCol)
+			if b.Debug {
+				log.Printf("Adding column %s %s %#x", colname, bulkCol.ColName, bulkCol.ti.TypeId)
+			}
+		} else {
+			return fmt.Errorf("Column %s does not exist in destination table %s", colname, b.tablename)
+		}
+	}
+
+	//create the bulk command
+
+	//columns definitions
+	var col_defs string
+	for i, col := range b.bulkColumns {
+		if i != 0 {
+			col_defs += ", "
+		}
+		col_defs += "[" + col.ColName + "] " + makeDecl(col.ti)
+	}
+
+	//options
+	var with_opts []string
+
+	if b.Options.CheckConstraints {
+		with_opts = append(with_opts, "CHECK_CONSTRAINTS")
+	}
+	if b.Options.FireTriggers {
+		with_opts = append(with_opts, "FIRE_TRIGGERS")
+	}
+	if b.Options.KeepNulls {
+		with_opts = append(with_opts, "KEEP_NULLS")
+	}
+	if b.Options.KilobytesPerBatch > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("KILOBYTES_PER_BATCH = %d", b.Options.KilobytesPerBatch))
+	}
+	if b.Options.RowsPerBatch > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("ROWS_PER_BATCH = %d", b.Options.RowsPerBatch))
+	}
+	if len(b.Options.Order) > 0 {
+		with_opts = append(with_opts, fmt.Sprintf("ORDER(%s)", strings.Join(b.Options.Order, ",")))
+	}
+	if b.Options.Tablock {
+		with_opts = append(with_opts, "TABLOCK")
+	}
+	var with_part string
+	if len(with_opts) > 0 {
+		with_part = fmt.Sprintf("WITH (%s)", strings.Join(with_opts, ","))
+	}
+
+	query := fmt.Sprintf("INSERT BULK %s (%s) %s", b.tablename, col_defs, with_part)
+
+	stmt, err := b.cn.Prepare(query)
+	if err != nil {
+		log.Fatal("Prepare failed:", err.Error())
+	}
+
+	if b.Debug {
+		log.Println(query)
+	}
+	_, err = stmt.Exec(nil)
+	if err != nil {
+		return err
+	}
+
+	b.headerSent = true
+
+	var buf = b.cn.sess.buf
+	buf.BeginPacket(packBulkLoadBCP)
+
+	//columns
+	columnMetadata := b.createColMetadata()
+	buf.Write(columnMetadata)
+
+	return nil
+}
+
+func (b *MssqlBulk) AddRow(row []interface{}) (err error) {
```

Comment:
This is an exported method on an exported type. It needs a comment in the form of:
// AddRow immediately/lazily writes the row to the bulk insert...